### PR TITLE
Rhmp

### PR DESCRIPTION
**Pull request by kmigielek**
red hat marketplace changes

### CODE REVIEW

File: pkg/resources/helper.go [hunk 1]
```diff
@@ -23,15 +23,19 @@ import (
 	"reflect"
 	"time"
 
+	"github.com/cloudflare/cfssl/log"
```
Comment (arturobrzut) [inline pkg/resources/helper.go:26]: can you removed this log ? cloudflare ?

File: pkg/resources/helper.go [hunk 2]
```diff
@@ -157,6 +163,15 @@ func UpdateServiceIfNeeded(reqLogger *logr.Logger, client c.Client, expectedServ
 	return reconcile.Result{}, nil
 }
 
+func UpdateServiceMonitor(reqLogger *logr.Logger, client c.Client, expected, found *monitoringv1.ServiceMonitor) (reconcile.Result, error) {
```
Comment (arturobrzut) [inline pkg/resources/helper.go:166]: can you move it to servicemonitor file

File: pkg/resources/helper.go [hunk 3]
```diff
@@ -199,6 +214,9 @@ func UpdateCacheClusterExtensions(client c.Reader) error {
 		IsRouteAPI = false
 	}
 
+	IsRHMP = checkRHMPPrereqs(client)
```
Comment (arturobrzut) [inline pkg/resources/helper.go:217]: why this is check?

File: pkg/resources/service/services.go [hunk 4]
```diff
@@ -58,6 +58,52 @@ func GetLicensingService(instance *operatorv1alpha1.IBMLicensing) *corev1.Servic
...
+					Protocol:   corev1.ProtocolTCP,
+				},
+			},
+			Selector: LabelsForSelector(instance),
+		},
+	}
+}
+
+func GetPrometheusServiceName() string {
+	return "license-service-prometheus"
```
Comment (arturobrzut) [inline pkg/resources/service/services.go:77]: set all constant name  in helpers

File: pkg/resources/service/services.go [hunk 5]
```diff
@@ -58,6 +58,52 @@ func GetLicensingService(instance *operatorv1alpha1.IBMLicensing) *corev1.Servic
...
+				},
+			},
+			Selector: LabelsForSelector(instance),
+		},
 	}
 }
+
+func getPrometheusLabels() map[string]string {
+	labels := make(map[string]string)
+	labels["release"] = "ibm-licensing-service-promethus"
```
Comment (arturobrzut) [inline pkg/resources/service/services.go:107]: set all constant name in helpers

File: pkg/controller/ibmlicensing/ibmlicensing_controller.go [hunk 6]
```diff
@@ -106,6 +107,17 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 		}
 	}
 
+	if res.IsRHMPEnabledAndInstalled(res.RHMPEnabled) {
+		// Watch for changes in prometheus related objects if rhmp is enabled
+		err = res.WatchForResources(reqLogger, &operatorv1alpha1.IBMLicensing{}, c, []res.ResourceObject{
+			&monitoringv1.ServiceMonitor{},
```
Comment (arturobrzut) [inline pkg/controller/ibmlicensing/ibmlicensing_controller.go:113]: add to csv and role yaml permision for service monitor

File: pkg/controller/ibmlicensing/ibmlicensing_controller.go [hunk 7]
```diff
@@ -106,6 +107,17 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
...
+			&extensionsv1.NetworkPolicy{},
```
Comment (arturobrzut) [inline pkg/controller/ibmlicensing/ibmlicensing_controller.go:114]: add to csv 1.4.X and role yaml permision for network policy

File: pkg/resources/helper.go [hunk 8]
```diff
@@ -208,3 +226,18 @@ func UpdateCacheClusterExtensions(client c.Reader) error {
 	}
 	return nil
 }
+
+func IsRHMPEnabledAndInstalled(rhmpEnabled bool) bool {
+	return rhmpEnabled && IsRHMP
+}
+
+func checkRHMPPrereqs(client c.Reader) bool {
+	prometheus := &apiextensionv1beta1.CustomResourceDefinition{}
```
Comment (arturobrzut) [inline pkg/resources/helper.go:235]: Reconsider checking all CRD vs checking the exact CRD 
it means or you add permission to list all CRD (in yaml role and csv)
or add permission to read / write service monitor or other CRD
send this to open slack channel Xteam or to me, @pgodowski @ab-ibm for agreement for it

File: pkg/resources/service/networking.go [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+	return &metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			"openshift.io/cluster-monitoring": "true",
+		},
+	}
+}
+func getNetworkPolicyFromPodSelector() *metav1.LabelSelector {
+	return &metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			"prometheus": "rhm-marketplaceconfig-meterbase",
```
Comment (arturobrzut) [inline pkg/resources/service/networking.go:92]: add it to helper as const or var

File: pkg/resources/service/networking.go [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+		MatchLabels: map[string]string{
+			"app": "ibm-licensing-service-instance",
+		},
+	}
+}
+
+func getNetworkPolicyFromNamespaceSelector() *metav1.LabelSelector {
+	return &metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			"openshift.io/cluster-monitoring": "true",
```
Comment (arturobrzut) [inline pkg/resources/service/networking.go:85]: add it to helper as const or var

File: pkg/resources/service/networking.go [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+				},
+			},
+		},
+	}
+}
+
+func getNetworkPolicyPodSelector() metav1.LabelSelector {
+	return metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			"app": "ibm-licensing-service-instance",
```
Comment (arturobrzut) [inline pkg/resources/service/networking.go:77]: add it to helper as const or var

File: pkg/resources/service/service_monitor.go [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+		Spec: monitoringv1.ServiceMonitorSpec{
+			Selector: metav1.LabelSelector{
+				MatchLabels: getPrometheusLabels(),
+			},
+			Endpoints: []monitoringv1.Endpoint{
+				{
+					BearerTokenSecret: corev1.SecretKeySelector{
+						Key: "",
+					},
+					Path:       "/metrics",
```
Comment (arturobrzut) [inline pkg/resources/service/service_monitor.go:48]: recheck if Path /metrics is default , if yes remove this line

File: pkg/resources/service/service_monitor.go [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+	return "license-service-service-monitor"
+}
+
+func GetServiceMonitor(instance *operatorv1alpha1.IBMLicensing) *monitoringv1.ServiceMonitor {
+	return &monitoringv1.ServiceMonitor{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      GetServiceMonitorName(),
+			Namespace: instance.Spec.InstanceNamespace,
+			Labels: map[string]string{
+				"marketplace.redhat.com/metering": "true",
```
Comment (arturobrzut) [inline pkg/resources/service/service_monitor.go:36]: add it to helper as const or var

File: pkg/resources/service/service_monitor.go [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+import (
+	monitoringv1 "github.com/coreos/prometheus-operator/pkg/apis/monitoring/v1"
+	operatorv1alpha1 "github.com/ibm/ibm-licensing-operator/pkg/apis/operator/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func GetServiceMonitorName() string {
+	return "license-service-service-monitor"
```
Comment (arturobrzut) [inline pkg/resources/service/service_monitor.go:27]: add it to helper as const or var

File: pkg/resources/helper.go (hunk 3 above)
Comment (arturobrzut) [reply pkg/resources/helper.go:217]: only for log?

File: pkg/controller/ibmlicensing/ibmlicensing_controller.go [hunk 15]
```diff
@@ -173,12 +185,18 @@ func (r *ReconcileIBMLicensing) Reconcile(req reconcile.Request) (reconcile.Resu
...
+		r.reconcileServices,
 		r.reconcileDeployment,
 		r.reconcileIngress,
 		r.reconcileRoute,
 	}
 
+	if res.IsRHMPEnabledAndInstalled(instance.Spec.IsRHMPEnabled()) {
+		reconcileFunctions = append(reconcileFunctions, r.reconcileServiceMonitor, r.reconcileNetworkPolicy)
+	} else {
+		reqLogger.Info("RHMP disabled or prometheus crd not present on cluster")
```
Comment (arturobrzut) [inline pkg/controller/ibmlicensing/ibmlicensing_controller.go:197]: if RHMP is disabled to not log in INFO 
if RHMP is enabled but CRD for RHMP does not exist then maybe you can add WARN log

File: pkg/controller/ibmlicensing/ibmlicensing_controller.go (hunk 7 above)
Comment (arturobrzut) [reply pkg/controller/ibmlicensing/ibmlicensing_controller.go:114]: and probably we do not want to watch NetworkPolicy

File: pkg/controller/ibmlicensing/ibmlicensing_controller.go (hunk 6 above)
Comment (arturobrzut) [reply pkg/controller/ibmlicensing/ibmlicensing_controller.go:113]: and we probably do not want to watch ServiceMonitor

**Review by arturobrzut — CHANGES_REQUESTED** [review]

File: pkg/resources/service/services.go [hunk 16]
```diff
@@ -58,6 +59,87 @@ func GetLicensingService(instance *operatorv1alpha1.IBMLicensing) *corev1.Servic
...
+			Selector: LabelsForSelector(instance),
+		},
+	}
+}
+
+func GetServiceMonitorName() string {
+	return "license-service-service-monitor"
+}
+
+func GetServiceMonitor(instance *operatorv1alpha1.IBMLicensing) *monitoringv1.ServiceMonitor {
```
[the remark this answers is not in the document]
Comment (kmigielek) [reply pkg/resources/service/services.go]: moved to new file

File: pkg/resources/service/services.go [hunk 17]
```diff
@@ -58,6 +59,87 @@ func GetLicensingService(instance *operatorv1alpha1.IBMLicensing) *corev1.Servic
...
+		Spec: monitoringv1.ServiceMonitorSpec{
+			Selector: metav1.LabelSelector{
+				MatchLabels: getPrometheusLabels(),
+			},
+			Endpoints: []monitoringv1.Endpoint{
+				{
+					BearerTokenSecret: corev1.SecretKeySelector{
+						Key: "",
+					},
+					Interval:   "15s",
```
[the remark this answers is not in the document]
Comment (kmigielek) [reply pkg/resources/service/services.go]: I removed it and it got changed to default

File: pkg/resources/service/services.go [hunk 18]
```diff
@@ -58,6 +59,87 @@ func GetLicensingService(instance *operatorv1alpha1.IBMLicensing) *corev1.Servic
...
+					Path:       "/metrics",
+					Scheme:     "https",
```
[the remark this answers is not in the document]
Comment (kmigielek) [reply pkg/resources/service/services.go]: There is new logic to manage scheme

File: pkg/resources/service/services.go [hunk 19]
```diff
@@ -58,6 +59,87 @@ func GetLicensingService(instance *operatorv1alpha1.IBMLicensing) *corev1.Servic
...
+					TargetPort: &prometheusTargetPort,
+					TLSConfig: &monitoringv1.TLSConfig{
+						InsecureSkipVerify: true,
```
[the remark this answers is not in the document]
Comment (kmigielek) [reply pkg/resources/service/services.go]: There is new logic to manage ssl

File: pkg/resources/helper.go (hunk 1 above)
Comment (kmigielek) [reply pkg/resources/helper.go]: removed

File: pkg/resources/service/networking.go (hunk 11 above)
Comment (kmigielek) [reply pkg/resources/service/networking.go]: moved to consts

File: pkg/resources/service/service_monitor.go (hunk 12 above)
Comment (kmigielek) [reply pkg/resources/service/service_monitor.go:46]: by default there is ""

File: pkg/resources/service/service_monitor.go (hunk 13 above)
Comment (kmigielek) [reply pkg/resources/service/service_monitor.go]: moved to consts

File: pkg/controller/ibmlicensing/ibmlicensing_controller.go (hunk 6 above)
Comment (kmigielek) [reply pkg/controller/ibmlicensing/ibmlicensing_controller.go]: service monitor is already added to role. Watcher removed

File: pkg/resources/service/service_monitor.go (hunk 14 above)
Comment (kmigielek) [reply pkg/resources/service/service_monitor.go]: moved to consts

File: pkg/controller/ibmlicensing/ibmlicensing_controller.go (hunk 15 above)
Comment (kmigielek) [reply pkg/controller/ibmlicensing/ibmlicensing_controller.go]: log removed

File: pkg/resources/helper.go (hunk 2 above)
Comment (kmigielek) [reply pkg/resources/helper.go:164]: I left it there because there is function which checks services as well

File: pkg/resources/helper.go (hunk 3 above)
Comment (kmigielek) [reply pkg/resources/helper.go]: It got change so function got removed

File: pkg/resources/service/services.go (hunk 4 above)
Comment (kmigielek) [reply pkg/resources/service/services.go]: moved to consts

File: pkg/controller/ibmlicensing/ibmlicensing_controller.go (hunk 7 above)
Comment (kmigielek) [reply pkg/controller/ibmlicensing/ibmlicensing_controller.go]: watcher removed and network policy permissions added

File: pkg/resources/helper.go (hunk 8 above)
Comment (kmigielek) [reply pkg/resources/helper.go]: it got changed so we check only if there is any instance of MarketplaceConfig

File: pkg/resources/service/networking.go (hunk 9 above)
Comment (kmigielek) [reply pkg/resources/service/networking.go]: moved to consts

File: pkg/resources/service/networking.go (hunk 10 above)
Comment (kmigielek) [reply pkg/resources/service/networking.go]: moved to consts

**Review by adamdyszy — APPROVED** [review]

**Review by arturobrzut — APPROVED** [review]

**Pull request merged**